[BE]스터디룸 그룹장 페이지 - API 구현

ChanYangYu opened the pull request:
## 관련 이슈

- closes #169

## 셀프 체크리스트

> 최소 1명 이상의 assign을 받아야만 Merge 가능합니다.

- [ ] Warning Message 발생 유/무
- [x] Coding Convention 준수 유/무
- [x] test 통과 여부

## 작업 요약
> 작업 내역 요약

- 그룹장 페이지 전체 API 구현
- 백엔드 API 테스트 코드 작성
## Preview

> 선택 사항

## PR 특이 사항

> PR을 볼 때 주의깊게 봐야하거나 말하고 싶은 점
> 백로그에 없지만 필요하여 구현한 기능 

- 현재 기능구현은 완료했고 테스트 코드 빠르게 작성중에 있습니다. 테스트 코드가 작성이 마무리될 때까지 기능 먼저 리뷰 해주시면 감사하겠습니다~

ChanYangYu committed:
[BE] feat: 그룹장 페이지 GET API 구현 


- 그룹장페이지 그룹정보  GET API 구현
- 그룹장페이지 그룹 참가 요청 리스트  GET API 구현
- issue (#169)

ChanYangYu committed:
[BE] refactor: 그룹장 페이지 컨트롤러 분리


- 기존 GroupController에서 GroupAdminController로 그룹장 페이지 로직 분리
- issue (#169)

ChanYangYu committed:
[BE] refactor: 그룹장 페이지 서비스 분리


- 기존 GroupService에서 GroupAdminService로 분리
- issue (#169)

ChanYangYu committed:
[BE] feat: 그룹 정보 수정 API 구현


- 그룹 이름, 그룹 시작/종료일, 그룹 소개 수정하는 API 구현
- issue (#169)

ChanYangYu committed:
[BE] feat: 그룹 참여 수락/거절 API 구현


- 그룹 참여 수락/거절 API 구현
  - groupService의 enroll 함수를 사용하기 위해 같은 서비스에 수락/거절 api를 구현
- 그룹 수락 API 로직에서 그룹에 등록하고 그룹 신청 상태를 변경하는 부분을 이후에 트랜잭션 처리로 바꿔줘야함
- issue (#169)

ChanYangYu committed:
[BE] feat: GroupApplyResponse 데이터 추가


- applyGroup Id 추가 ( 수락/거절 시 필요)
- issue (#169)

ChanYangYu committed:
[BE] feat: 컨트롤러 isLoggedIn 미들웨어 추가


- 로직 처리전 로그인 여부 체크 하도록 변경
- issue (#169)

ChanYangYu committed:
[BE] feat: 그룹참가 요청 정보 깃허브 아이디 추가


- 프로필 리다이렉트를 위한 깃허브 아이디 정보 추가
- issue (#169)

ChanYangYu committed:
[BE] feat: 로그인한 유저가 그룹장인지 확인로직 추가


- 로그인된 유저가 그룹장인지 확인 로직 추가
- issue (#169)

ChanYangYu committed:
[BE] feat: 그룹참가 수락/거절시 그룹장 체크 로직 추가


- 그룹 참가 수락/거절 API 수행시, 수락/거절을 요청한 사람이 해당 요청의 그룹장인지 체크하는 로직 추가
- issue (#169)

ChanYangYu committed:
[BE] rename: 그룹설정 API 이름 변경


- group/admin -> group-owner 
- issue (#169)

ChanYangYu committed:
[BE] refactor: 그룹 참여 수락/거절 API 이동


- groupOwnerService로 이동 시키고 이 서비스에 groupService 주입
- issue (#169)

Donggggg commented on server/src/domains/group/api/GroupOwnerController.ts:
```diff
@@ -0,0 +1,94 @@
+  @ResponseSchema(SimpleGroupInfoResponse)
+  @UseBefore(isLoggedIn)
+  @Get('/:gid')
+  public async getGroupInfo(@Session() session: any, @Param('gid') groupId: number) {
+    await this.groupOwnerService.checkIsGroupOwner(groupId, session.user.id);
+    return await this.groupOwnerService.getGroupInfoByGroupId(groupId);
```
파라미터 유효성 검증이 필요할 것 같아요!

ChanYangYu replied:
추가했습니다~

Donggggg commented on server/src/domains/group/repository/ApplyGroupRepository.ts:
```diff
@@ -14,4 +14,20 @@ export class ApplyGroupRepository extends Repository<ApplyGroup> {
   }
-}
\ No newline at end of file
+  public findApplyListByGroupId(gid: number) {
+    return this.createQueryBuilder('apply_group')
+      .innerJoinAndSelect('apply_group.user', 'user')
+      .where('apply_group.groupId = :gid', { gid })
+      .andWhere('state = "PENDING"')
+      .orderBy('apply_group.createdAt', 'ASC')
+      .getMany();
+  }
+
+  public findApplyData(id: number, ownerId: number) {
+    return this.createQueryBuilder('apply_group')
+      .innerJoinAndSelect('apply_group.group', 'group')
+      .where('apply_group.id = :id', { id })
+      .andWhere('group.ownerId = :ownerId', { ownerId })
+      .getOne();
+  }
+}
```
레포지토리 문법으로 변경할 수 있는 쿼리 같습니다!

ChanYangYu replied:
@Donggggg  적용완료했습니다!

Donggggg commented on server/src/domains/group/service/GroupOwnerService.ts:
```diff
@@ -0,0 +1,81 @@
+    private readonly groupService: GroupService,
+  ) {}
+
+  public async checkIsGroupOwner(gid: number, userId: number): Promise<void> {
+    const { ownerId } = await this.groupRepository.findOneOrFailByGroupId(gid);
+
+    if(ownerId !== userId){
+      throw new NotAuthorizedError();
+    }
+  }
```
groupService에 해당 역할을 하는 서비스가 존재하는데 재사용하면 좋을 것 같아요!

ChanYangYu replied:
해당 코드 재사용하도록 코드 변경했습니다!

Donggggg commented on server/src/domains/group/service/GroupOwnerService.ts:
```diff
@@ -0,0 +1,81 @@
+  public async acceptRequest(applyId: number, ownerId: number) {
+    const applyGroup = await this.applyGroupRepository.findApplyData(applyId, ownerId);
+
+    if (applyGroup === undefined) {
+      throw new NotAuthorizedError();
+    }
```
`!applyGroup` 은 어떤가요?!

Donggggg replied:
75번 라인도요!

ChanYangYu replied:
피드백 적용완료했습니다!

ChanYangYu committed:
[BE] feat: param 타입체크, 중복 코드 제거


- checkIsGroupOwner 중복코드 제거 및 재사용
- param 타입 체크
- issue (#169)

ChanYangYu committed:
[BE] refactor: 쿼리빌더 제거 및 Param 수정


- 컨트롤러 Param -> Params
- 레포지토리 네이밍 변경 및 쿼리빌더 제거
- issue (#169)

ChanYangYu committed:
[BE] test: API 테스트 코드 작성


- 공통적으로 그룹장인지 여부 체크를 실행하고 추가적으로 GET API의 경우 제대로된 값을 가져오는지 body 값체크 수행
- issue (#169)

ChanYangYu committed:
[BE] feat: 그룹장 페이지 시드 추가 및 테스트 코드 적용


- issue (#169)

who-is-hu commented on server/src/domains/group/service/GroupOwnerService.ts:
```diff
@@ -0,0 +1,74 @@
+  ) {}
+
+  public async getGroupInfoByGroupId(gid: number): Promise<SimpleGroupInfoResponse> {
+    const { name, intro, startAt, endAt } = await this.groupRepository.findOneOrFailByGroupId(gid);
```
이거 좀 고민이긴한데 이렇게하면 편하긴한데 dto에 준일님이 말해줬던 assgin 써서 옮기면
```
const group = await this.groupRepository.findOneOrFailByGroupId(gid);
return SimpleGroupInfoResponse.from(group);
```
하면 훨씬 읽기 좋긴하거든요 만약 여러곳에서 이 dto쓴다하면 `{ name, intro, startAt, endAt }` 이코드가 서비스 로직에 계속 등장해서 안좋을 것 같아요

ChanYangYu replied:
오 적용시켰습니다!

who-is-hu commented on server/src/domains/group/repository/ApplyGroupRepository.ts:
```diff
@@ -14,4 +14,24 @@ export class ApplyGroupRepository extends Repository<ApplyGroup> {
   }
-}
\ No newline at end of file
+  public findByGroupIdAndState(gid: number, state: string) {
```
findAllBy

who-is-hu commented on server/src/domains/group/service/GroupOwnerService.ts:
```diff
@@ -0,0 +1,74 @@
+  }
+
+  public async getApplyListByGroupId(gid: number): Promise<GroupApplyResponse[]> {
+    const applyGroupList = await this.applyGroupRepository.findByGroupIdAndState(gid, 'PENDING');
```
'PENDING' 대신 만들어둔 enum 을 쓰는게 좋을 것 같습니다.

ChanYangYu replied:
예스! 적용완료했습니다!

who-is-hu commented on server/src/domains/group/service/GroupOwnerService.ts:
```diff
@@ -0,0 +1,74 @@
+  public async getApplyListByGroupId(gid: number): Promise<GroupApplyResponse[]> {
+    const applyGroupList = await this.applyGroupRepository.findByGroupIdAndState(gid, 'PENDING');
+    return applyGroupList.map(({ id, createdAt, user }) => {
+      return {
```
이것도 dto에 from 메서드 만들고 하면 한줄로 줄여서 서비스 로직 간단하게 할 수 있을 것 같습니다.
```
return applyGroupList.map(group => GroupApplyResponse.from(group));
```

ChanYangYu replied:
이부분도 적용완료했습니다!

who-is-hu commented on server/src/domains/group/service/GroupOwnerService.ts:
```diff
@@ -0,0 +1,74 @@
+    if (!applyGroup) {
+      throw new NotAuthorizedError();
+    }
+    applyGroup.state = ApplyGroupState.ACCEPTED;
```
이 줄도 트랜잭션에 포함되어야 할 것 같습니다.
`applyGroup.state = ApplyGroupState.ACCEPTED;`

ChanYangYu replied:
아 save를 나중에 하는데 바로 디비에 적용이 되는건가요??

who-is-hu replied:
ㅇㅎ 그러네요 리팩토링때 함 실험해보고 정합시당

who-is-hu commented on server/src/domains/group/service/GroupOwnerService.ts:
```diff
@@ -0,0 +1,74 @@
+    return this.groupRepository.update({ id: gid }, { intro });
+  }
+  public async acceptRequest(applyId: number, ownerId: number) {
+    const applyGroup = await this.applyGroupRepository.findOneByIdAndOwnerId(applyId, ownerId);
```
GroupApply를 찾는거면 이미 pk가 있는데 ownerId가 같이 필요한 이유가 있을까요?

ChanYangYu replied:
아 이부분은 로직과 관련있는데 그냥 applyId만 비교하게 되면 
로그인한 다른 유저가 applyId를 알고있으면 
`api/group-owner/accept/1` 요청을 보내 그룹 참여 수락으로 바꿔버리는 문제가 있어
로그인 한 유저가 그룹장인지 체크를 하려고 저렇게 로직을 처리했습니다!

who-is-hu replied:
그래도 ApplyGroup을 찾는 함수에 ownerId 가 들어가는 이유는 안되는것같아요, 찾는건 이미 pk하나로만 식별해서 찾을 수 있습니다. 조인을 해서 가져오고 서비스 로직에서 검사를 하는게 맞는것 같아요.

이건 참고로 조인한 값의 어트리뷰트를 찾는경우 네이밍을 `findOneByIdAndGroupOwnerId` 이런식으로 한답니다.

who-is-hu commented on server/src/domains/group/service/GroupOwnerService.ts:
```diff
@@ -0,0 +1,74 @@
+  public async acceptRequest(applyId: number, ownerId: number) {
+    const applyGroup = await this.applyGroupRepository.findOneByIdAndOwnerId(applyId, ownerId);
+
+    if (!applyGroup) {
```
findOrFail 쓰면 안되는 곳인가요?

ChanYangYu replied:
앞의 이유 때문에 이렇게 처리했습니다!

who-is-hu commented on server/src/domains/group/repository/GroupRepository.ts:
```diff
@@ -87,4 +87,8 @@ export class GroupRepository extends Repository<Group> {
     });
   }
+
+  public findOneOrFailByGroupId(groupId: number){
```
pk 조회니까 그냥 findOneorFailById 아닐까요

ChanYangYu replied:
오 그러네요!

who-is-hu commented on server/src/domains/group/repository/ApplyGroupRepository.ts:
```diff
@@ -14,4 +14,24 @@ export class ApplyGroupRepository extends Repository<ApplyGroup> {
+    });
+  }
+
+  public findOneByIdAndOwnerId(id: number, ownerId: number) {
```
서비스 로직에도 코멘트 달았는데 ownerId 가 왜필요한지 모르겠습니다.

who-is-hu commented on server/src/domains/group/dto/UpdateGroupDto.ts:
```diff
@@ -0,0 +1,24 @@
+  @IsNumber()
+  gid: number;
+  @IsString()
+  name: string;
```
글자수 제한 박아야하지않나요?

ChanYangYu replied:
적용완료했습니다!

who-is-hu commented on server/src/domains/group/dto/UpdateGroupDto.ts:
```diff
@@ -0,0 +1,24 @@
+  @IsNumber()
+  gid: number;
+  @IsString()
+  intro: string;
```
글자수 제한 박아야하지않나요?

ChanYangYu replied:
적용완료했습니다!

who-is-hu commented on server/src/domains/group/api/GroupOwnerController.ts:
```diff
@@ -0,0 +1,100 @@
+  @UseBefore(isLoggedIn)
+  @Patch('/decline/:aid')
+  public async declineApply(@Session() session: any, @Params() { aid }: ApplyParam) {
+    console.log(aid);
```
delete!

ChanYangYu replied:
오 이거 고민했었는데 바꿔야겠네요!

who-is-hu replied:
아녀 console.log지워주세요  ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ

ChanYangYu replied:
앗ㅋㅋㅋㅋ 삭제하겠습니다!

who-is-hu commented on server/src/domains/group/dto/ApplyGroupDto.ts:
```diff
@@ -6,3 +6,8 @@ export class ApplyGroupDto {
   userId: number;
 }
+
+export class ApplyParam {
```
위의 ApplyGroupDto와 무조건 같이 쓰이는 Param validation이 아니라
별도의 파일로 분리하는게 좋을거 같습니다

ChanYangYu replied:
분리했습니다!

who-is-hu commented on server/src/domains/group/dto/UpdateGroupDto.ts:
```diff
@@ -0,0 +1,24 @@
+import { IsDateString, IsNumber, IsString } from "class-validator";
+
+export class UpdateGroupNameDto {
```
파일 분리하고 폴더로 묶는게 날것같습니다.

ChanYangYu replied:
넵!

who-is-hu commented on server/src/domains/group/api/GroupOwnerController.ts:
```diff
@@ -0,0 +1,100 @@
+import { GroupOwnerService } from '../service/GroupOwnerService';
+import { UpdateGroupDateDto, UpdateGroupIntroDto, UpdateGroupNameDto } from '../dto/UpdateGroupDto';
+import { GroupService } from '../service/GroupService';
+import { GroupParam } from '../dto/groupDto';
```
예도 불필요하게 묶여있는거 같은데 주말에 분리합시당

ChanYangYu replied:
넵 리팩토링이 필요해보이네요!

who-is-hu commented:
테스트 여부에 체크가 안되어있는데 로컬에서 테스트 돌리면 패스 못하나요?
그럼 통과하게 고쳐서 올리면 좋을 것 같습니다.

ChanYangYu committed:
[BE] refactor: dto 내부 from 메소드 구현 및 enum 적용


- issue (#169)

ChanYangYu committed:
[BE] refactor: dto 분리


- issue (#169)

ChanYangYu committed:
[BE] remove: console.log제거


- issue (#169)

ChanYangYu committed:
[BE] refactor: eager 옵션 로직 적용


- 레포지토리 로직 간소화 및 서비스 로직에서 비교처리 하도록 변경
- issue (#169)

who-is-hu commented on server/src/domains/group/repository/GroupRepository.ts:
```diff
@@ -87,4 +87,8 @@ export class GroupRepository extends Repository<Group> {
   }
+
+  public findOneOrFailById(id: number) {
+    return this.findOneOrFail({ id });
```
findOneOrFail(id) 아닌가요?

ChanYangYu committed:
[BE] feat: 레포지토리 코드 간결화


- this.findOneOrFail({ where: { id } }); ->  this.findOneOrFail(id)

yuchanleeme reviewed (approved):
다른분들이 의견을 많이 남겨주셔서 저는 수정된 코드만 봤을 때 괜찮다고 생각합니다!! 굿굿!!
